### ISSUE TITLE
Create a way to notify the apps when static-file-cache is flushed

### ISSUE THREAD

**Issue by alshakero** [Starcounter/Home#439]
### What we need 

Currently, Starcounter serves static asset files from `/sys` folder in every running app. And when the same file is requested more than once, it is served from some sort of cache. 

I assume that Starcounter watches the files that it caches, and flushes the cache whenever they change. We need a way to subscribe and be notified when this event happens.

If I'm wrong and Starcounter does not watch the files, it's not an issue, all we need is an event that fires whenever any file in one of the served `sys` folders is modified. The way you implement that is totally up to you. 

### Motivation

This is important for Service Worker cache invalidation.  We'll soon release Service Worker support (read: a cache mechanism on the client side — the browser) and when the files changes on the disk, we'll need to flush the browser cache so the disk and the browser become eventually-consistent. 

### Benefit

Our apps will be much faster, especially on mobile phones.

//cc @warpech @alemoi

**Comment by alshakero** [comment]
Please put this on hold until we discuss it in Gdansk.

**Comment by warpech** [comment]
I can't find the discussion now, but I remember that @miyconst wrote to me on Slack that it is an anti-pattern to implement this, because it will be impossible to support in multi process Nova, which is designed NOT to have a common file server.

@miyconst pls confirm.

**Comment by miyconst** [comment]
@warpech, unfortunately, this discussion was only held offline, as far as I remember. 

There are a few major concerns with this approach, and I don't think it shall be part of any app at all. If we implement it, it shall be a part of the Polyjuice platform.

- Multi-host, as you mentioned, different apps will use different web server.
- Way to many points of failure:
  - What if the app is not running during the update?
  - Race condition between file update and file update notifications.
  - What if multiple apps serve the same HTML file from the `/sys/` folder?
  - Such cache invalidation should not be performed in production at all.

**Comment by tomalec** [comment]
>     * What if the app is not running during the update?

I think actual need is to notify Starcounter's app-shell. So if we will integrate SCF/`app-shell.html` to the Polyjuice installation, it will not be a case.

>     * Race condition between file update and file update notifications.

Could you describe such race condition?

>     * What if multiple apps serve the same HTML file from the `/sys/` folder?

I think for our needs a solution where notification will be send for a chnage file, that is overwritten by other file which was not change, is fine. That means we will invalidate the cache even though we didn't have to. Not perfect, but better than serving the old file if it actually changed.

>     * Such cache invalidation should not be performed in production at all.

Why? I think it should. In scenarios like:
1. StarcounterClientFiles update
  1. End user opens the browser and visits the solution, that's running, let's say Palindrom 1.2.3
  2. Solution owner decides to update their stack of apps
  3. End user re-visits the solution
2. Composition update
  1. End user opens the browser and visits the solution, that's running, let's say Palindrom 1.2.3
  2. Solution owner decides to update their stack of apps
  3. End user re-visits the solution

**Comment by miyconst** [comment]
@tomalec can we have an online or offline meeting about it again? I strongly believe that this feature should not be implemented, since it's extremely error prone and requires way too much investment.

Some answers to the questions:

- Integrating app shell into Starcounter won't help because there is no guarantee, that Starcounter will be running at the time of the update.
- Race condition will happen when there are multiple subscribes and multiple changes, especially when subscribes are also performing some file changes. This means all subscribes shall be thread-safe, and be able to schedule notifications queue, and don't kill each other with locking files and similar issues.
- When an app upgrade happens, apps version shall be updated accordingly, the version shall be included into the URL / URI or any other caching mechanism, to make sure that cache is automatically invalidated with an app upgrade.

**Comment by tomalec** [comment]
> can we have an online or offline meeting about it again?

Sure, I'd be happy to. Please, propose a date, maybe somewhere in next week or two, as I guess you are pretty busy right now.

**Comment by warpech** [comment]
The meeting is scheduled on 26 Feb.

**Comment by warpech** [comment]
@miyconst, @alshakero, @tomalec and me agreed today that this is needed and useful for any projects that have web GUIs. The more components, the bigger the benefit.

The decisions made:

1. cache key should regenerate on server start, in case a URI changed when the server was stopped
   - we have an implementation of this point as proposed changes to `PartialToStandaloneHtmlProvider`, commited as a fork of this middleware in https://github.com/Starcounter/UniformDocs/pull/357.
2. cache key should regenerate on any app start, in case this app replaces the content at the URI
   - not prototyped yet
3. cache key should regenerate whenever any static file is modified (this requires a new API in Level1 that will be consumed in `PartialJsonToStandaloneHtmlProvider.cs`)
   - not prototyped yet
   - @alshakero suggests that we plug into the code in Level1 that replaces the Starcounter memory cache of a static file, whenever the file watcher detects a change in the file on disk
   - @miyconst suggests, that there might be a simpler initial solution that can be implemented fully within `PartialJsonToStandaloneHtmlProvider.cs`. The solution is to obtain the list of static file directories at the time when the app registers to use the middleware and set up separate file watchers inside `PartialJsonToStandaloneHtmlProvider`. The list of static file folders can be obtained from:
      - `Starcounter.Application.Current.ResourceDirectories`
      - or, if the above is empty (or null?), from `Starcounter.Application.Current.WorkingDirectory + "/wwwroot"`
4. a REST API to flush the keys on demand

We agree that it would be rational for @alemoi to work on it. Otherwise @alshakero or @warpech can probably substiture for it but Alexey will do it better.

**Comment by warpech** [comment]
@miyconst, @alemoi please inform if @alemoi can proceed with "right" implementation, or should my team provide an ad-hoc implementation with a new file watcher in `PartialJsonToStandaloneHtmlProvider`

**Comment by alemoi** [comment]
I will do it, just negotiated with @miyconst

**Comment by alemoi** [comment]
As I understand a cache key is assigned to a static resource, and is retrievable by resource URI (i.e. `request.Uri`). So my questions are: 

> 1. cache key should regenerate on server start, in case a URI changed when the server was stopped

When SC server starts, there is nothing in file cache, because there is no file cache even created yet, there is no codehost even started. Can you explain please?

> 2. Cache key should regenerate on any app start, in case this app replaces the content at the URI

Which cache key? How to determine which keys should be regenerated? Can the whole cache be flushed or it will affect other apps, and thus shouldn't? By the way since codehost restarts when app is restarted - file cache is flushed automatically and completely.

> 3. Cache key should regenerate whenever any static file is modified

This is doable since resource is trackable, but previous questions are unclear to me. CC @warpech

**Comment by warpech** [comment]
@alemoi please note that it is not the server-side cache that we are invalidating here. It is a client-side cache that we need to invalidate. The client only knows what the server tells him, and that includes the identifier for cacheable resources ("Service Worker cache key").

> As I understand a cache key is assigned to a static resource, and is retrievable by resource URI (i.e. `request.Uri`).

Not actually, the single Service Worker cache key represents the collective version of all static files served by the static file server. If any file was changed (or we don't know if it was changed, because we were not observing), the cache key needs to be regenerated.

You can see the usage of the SW cache key here: https://github.com/Starcounter/UniformDocs/pull/357/files (see line 12 of the first file in the review)

> When SC server starts, there is nothing in file cache, because there is no file cache even created yet, there is no codehost even started. Can you explain please?

When SC server starts, there might be a Service Worker cache already populated in a client. So you need to regenerate the key, so that the client realizes that his cache is outdated.

**Comment by alemoi** [comment]
I have chatted with @warpech and understood the situation: on the server inside `PartialJsonToStandaloneHtmlProvider` you need to monitor only one file `service-workers.js` and that's it. I have checked our static file server implementation, and it does not have a direct functionality to monitor individual files, I would have to create it just to support this case, which I don't think is right. My proposal: I have added the API to get physical static file path by its URI via `Handle.Handle.GetFilePathFromUri(uri)` and here is the working example of how I monitor the changes on file (including renaming and deleting):
```cs
var filePath = Handle.GetFilePathFromUri(uri);
WatchChange(filePath, FileHasChanged, FileIsRenamed);

...

void FileIsRenamed(object sender, RenamedEventArgs e)
{
    string fileSignature = e.OldFullPath.ToLower();
}

 void FileHasChanged(object sender, FileSystemEventArgs e)
{
    string fileSignature = e.FullPath.ToLower();
}

void WatchChange(string filePath, FileSystemEventHandler fileChanged, RenamedEventHandler fileRenamed)
{
    FileSystemWatcher fsw;
    fsw = new FileSystemWatcher();
    fsw.InternalBufferSize = 64 * 1024;
    fsw.Path = Path.GetDirectoryName(filePath);
    fsw.Filter = Path.GetFileName(filePath);
    fsw.IncludeSubdirectories = false;
    fsw.Changed += fileChanged;
    fsw.Deleted += fileChanged;
    fsw.Renamed += fileRenamed;
    fsw.EnableRaisingEvents = true;
}

```

Let me know if that is sufficient @warpech

**Comment by tomalec** [comment]
> you need to monitor only one file `service-workers.js` and that's it

I think we need to **monitor all** static files, or at least a white list of files. Then `PartialJsonToStandaloneHtmlProvider` **uses** a notification to change `service-workers.js`.

**Comment by alemoi** [comment]
I asked @warpech and he confirmed its only one file `service-workers.js`, lets see what he says here.

**Comment by alemoi** [comment]
Ok, now I see. Will try to retrieve all folders from static file server for monitoring.

**Comment by eriohl** [comment]
Regardless of the number of files it seems to me that adding a file watch on a static file seems not like a something we should be seeing a lot in application code right? So then it will not be supported by a public method in `Starcounter.Handle`.

**Comment by tomalec** [comment]
We need that for the app shell that initializes frontend framework. 
So that's not the app's code, unless the app want's to overwrite the app shell delivered with the installation.

**Comment by alemoi** [comment]
As I wrote, I have checked the static web server and its mechanisms are totally different from the desired. For example, static file server sets "monitors" on the file system only when the file is accessed (it takes file folder to monitor): https://github.com/Starcounter/level1/blob/develop/src/Starcounter.Hypermedia/StaticWebServer.File.cs#L222 Maybe its done this way because monitoring folders is expensive (not sure). In our case, a middleware wants to monitor certain folders (not all folders from all apps). I don't think it should be done by the static file server with its limited functionality as it is, just to support this specific case. I suppose it should be done by the middleware that wants this or our static file server should be rewritten. At the moment static file server provides registered directory paths or as mentioned above converts a file URI into file path. If it should be provided by our static file server, then I will do, but I need an advice CC @miyconst @eriohl (since you decided to participate)

**Comment by miyconst** [comment]
If a file was not requested, the it was not cached by the browser, then it makes no sense to flash cash when it's changed.

Thus, I suggest to add monitored files / folders the same way Static Files server does it. Flush the cache when a change in one of the monitored files / folders happen.

**Comment by eriohl** [comment]
My point was that making `Handle.GetFilePathFromUri` public doesn't seem very appropriate (especially since it didn't even solve the problem). Not I think is having application code working with files.

Web server static files isn't even supposed to be served by the worker processes and you certainly shouldn't assume it is.

**Comment by eriohl** [comment]
My point was that making `Handle.GetFilePathFromUri` public doesn't seem very appropriate (especially since it didn't even solve the problem). Nor I think is having application code work with static web server files. Likely the worker process won't even be able to access them (as the built-in web server won't be the one serving them to the client).

**Comment by eriohl** [comment]
Well: We plan to replace built-in static web server as soon as possible and the worker process is not suppose to serve the static web server files anyway (we only added this for development convenience). So unless this is an immediate concern it seems like not something we should be doing at all right now in my opinion. (But I only got involved because this `Handle.GetFilePathFromUri` being made public is something that I felt needed to be reverted.)

**Comment by eriohl** [comment]
> ... So unless this is an immediate concern it seems like not something we should be doing at all right now in my opinion. ...

Later learned that is is in fact of immediate concern so never mind.

**Comment by alemoi** [comment]
After discussing with @mtornwall since its needed immediately today I'm adding a solution.

**Comment by alemoi** [comment]
I have added a temporary solution. Lets see how it goes (reference to assembly `Starcounter.Apps.JsonPatch` should be changed to `Starcounter.App.Server`):
```cs
static void FileRenamed(object sender, RenamedEventArgs e)
{
    string fileSignature = e.OldFullPath.ToLower();
}

static void FileChanged(object sender, FileSystemEventArgs e)
{
    string fileSignature = e.FullPath.ToLower();
}

static void FileDeleted(object sender, FileSystemEventArgs e)
{
    string fileSignature = e.FullPath.ToLower();
}
...
AppsBootstrapper.WatchDirectories(FileDeleted, FileChanged, FileRenamed);
```

**Comment by warpech** [comment]
I didn't try to run it yet, but I have few comments that I discussed with @alemoi on a Slack call:

1. We don't need to three distinct events like `FileDeleted`, `FileChanged`, `FileRenamed`. One event `ResourceChanged` is sufficient. 
2. The event should have a property that contains the resource URI. The file path is irrelevant for this event.
2. We should abstract away the fact that static resources come from observing files in directories. For the user of this method, we are observing static resources, which do not have to be files. It is better to rename the method `WatchDirectories` to `WatchResources`.
3. An app can have dynamic handlers to generate static resources that are not files on disk. Therefore, in addition to using `FileSystemWatcher` as the source of information, we could add a method `AppsBootstrapper.NotifyResourceChange(uri)` that an app can use to inform `AppsBootstrapper` that a resource was changed.

**Comment by warpech** [comment]
I have tried to use the API in https://github.com/Starcounter/UniformDocs/pull/357.

I have some problems, please check my last comment there. The `WatchResources` callback is not called.

**Comment by warpech** [comment]
The above problem seems to be gone. Thank you @alemoi for the nice implementation. Things still needed:

- [ ] documentation

Anything else?

**Comment by alemoi** [comment]
Since this is experimental, lets wait with documentation till confirmation point.

**Comment by miyconst** [comment]
Closing as won't fix in Starcounter 2.4 and not applicable in Starcounter 3.0.
